CP-18185 new plugin discovery mechanism

lindig opened the pull request:
I'm inviting comments for the implementation of a plugin discovery mechanism that is part of PVS. 
- For this to build I'd need to change the corresponding spec file such that `ocaml-inofify` is included. However, do we use branches for the spec file repositories?
- I still need to test this. (inotify does not work inside Docker.)
- What should happen to the existing registration mechanism? One way to solve this is to keep it and to have the new mechanism look for plugins in a different directory than `/dev/shm/metrics` (`Rrdd_server.Plugin.base_path`).

lindig committed:
CP-18185 don't enable coverage analysis by default for rrdd_libs

Coverage analysis is explicitly activated using a Makefile target. I
suspect this is a bug from an oversight.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
CP-18185 add utility rrd_discover for plugin discovery

The tool implements inotify-based plugin discovery and served as a
test bed for experiments.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
CP-18185 implement new plugin discovery

Plugins are discovered in the file system by a thread that watches a
directory in the file system.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
CP-18185 update auto-generated files

This commit contains changes for auto-generated files caused by changed
from previous commits.

Signed-off-by: Christian Lindig <lindig@gmail.com>

robhoes commented on rrdd/xcp_rrdd.ml:
```diff
@@ -615,6 +695,13 @@ let _ =
+	let () = 
+		try Discover.start () |> ignore
+		with e -> error 
+			"RRD plugin discovery thread failed: %s"
```
Let's restart the thread on failure (after a short delay).

robhoes commented:
Looks pretty good to me so far!

lindig committed:
CP-18185 restart plugin discovery after failure

The plugin thread gets restarted after 10 seconds after its failure.=

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
CP-18185 just use basename of file as UID when registering plugin

The Plugin module expects the basename of a file to be used as UID.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
Revert "CP-18185 restart plugin discovery after failure"

This reverts commit b9edd48e19be73d1b87fd148f15009064d5ceaff.

The solution to restarting the thread is fundamentally broken.

lindig committed:
CP-17268 fix thread startup for discovery thread

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig commented:
The code works now when built against `ocaml-inotify 2.3` and tested on a DT box. A remaining problem is that RRD files created by `rrd-client-lib` can't be processed and lead to exceptions - most likely because of format incompatibilities. I've submitted a pull request for `ocaml-inotify 2.3`. To see the full logs, one needs to edit `/ect/xcp-rrdd.conf`.

lindig committed:
CP-18185 cleanup: log levels, dead code removed

When a plugin is registered or un-registered, this is now logged on
level info. One unused let binding removed.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig commented:
The code now works and successfully registers and un-registers plugins on a DT machine. However, the mechanism is still a bit fragile for a different reason: because data is immediately read when a plugin is registered, a plugin cannot create the file and write content only later. It has to provide valid content as it creates the file. This is currently not the case for files created by rrd-client-lib. This library creates a file and updates it later, leading to read errors in xcp-rrdd.

lindig committed:
CP-18185 watch plugin directory also for move events

So far the directory was only watched for create and delete events and
would miss files that are moved in our out of it, or being renamed. This
commit rectifies this.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig commented:
The remaining problem of plugin registration being fragile is addressed by changes on the client side at https://github.com/lindig/rrd-client-lib-1/tree/rrdd-protocol. The client now writes a file of constant size and makes sure the content is valid.

lindig committed:
CP-18185 ignore *.tmp as RRD files, restart discovery loop after failure

We now ignore some files in the watched directory to give Plugins the
opportunity to use an atomic rename(2) for submitting data.

When plugin discovery fails for any reason, it is resumed after 10
seconds.

Signed-off-by: Christian Lindig <lindig@gmail.com>

jonludlam commented:
There's a lot of code duplicated between the standalone util and rrdd - if we're intending to keep the stand-alone tool around we should dedup, and if not, we should rm it.

lindig commented:
That is a fair point. I was using it for experiments with the inotify mechanism. It is more a tool than an integrated part.

lindig committed:
CP-18185 remove rrd_discover.ml

rrd_discover is a stand-alone tool for early experiments and no longer
needed.

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
CP-18185 update auto-generated files from changes to _oasis

Commit 9e4bc36 changed the _oasis file. This requires updating
auto-generated code.

Signed-off-by: Christian Lindig <lindig@gmail.com>

jonludlam commented on rrdd/xcp_rrdd.ml:
```diff
@@ -538,6 +538,121 @@ let monitor_loop () =
+		debug "RRD plugin - scanning %s" dir;
+		Sys.readdir dir
+		|> Array.to_list
+		|> List.iter register
```
Do you need a filter by is_valid here?

lindig replied:
Yes - I should add it. The expectation is that these tmp files are very short lived and don't exist at startup time. But they could exist when we recover from a crash. In any case, we should always ignore tmp files.

lindig committed:
CP-18185 filter out *.tmp files when scanning directory at startup

This commit brings the behavior of discover() in line with that of
watch().

Signed-off-by: Christian Lindig <lindig@gmail.com>

jonludlam commented:
Cool, looks good now.
